### PR TITLE
chore(website): [playground] allow to choose file extensions

### PR DESCRIPTION
**Pull request by armano2**
## PR Checklist

- [x] Addresses an existing open issue: fixes #000
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

> extracted from #6656 

This change replaces `jsx` checkbox with dropdown of supported file extensions

image

ref #5449

### CODE REVIEW

File: packages/website/src/components/editor/LoadedEditor.tsx [hunk 1]
```diff
@@ -28,12 +28,25 @@ export interface LoadedEditorProps extends CommonEditorProps {
...
+    return 'typescript';
+  }
+  if (/\.[mc]?js(x)?$/.test(file)) {
+    return 'javascript';
+  }
+  if (/\.(json|eslintrc)$/.test(file)) {
+    return 'json';
+  }
+  return 'plaintext';
+}
```
Comment (bradzacher) [inline packages/website/src/components/editor/LoadedEditor.tsx:42]: could this use this function instead?

https://github.com/typescript-eslint/typescript-eslint/blob/399b3b1263665288316be6faff759cf514e24b1b/packages/typescript-estree/src/create-program/getScriptKind.ts#L4-L34

File: packages/website/src/components/config.ts [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
+  'ts',
+  'tsx',
+  'js',
+  'jsx',
+  'd.ts',
+  'cjs',
+  'mjs',
+  'cts',
+  'mts',
 ];
```
Comment (bradzacher) [inline packages/website/src/components/config.ts:20]: could this be `Object.values(ts.Extension)` instead?

File: packages/website/src/components/types.ts [hunk 3]
```diff
@@ -14,14 +14,27 @@ export interface RuleDetails {
...
+export type ConfigFileType =
+  | 'ts'
+  | 'tsx'
+  | 'js'
+  | 'jsx'
+  | 'd.ts'
+  | 'cjs'
+  | 'mjs'
+  | 'cts'
+  | 'mts';
```
Comment (bradzacher) [inline packages/website/src/components/types.ts:26]: could this be `ts.Extension` instead?

File: packages/website/src/components/Playground.tsx [hunk 4]
```diff
@@ -50,7 +50,7 @@ function rangeReducer<T extends SelectedRange | null>(
 }
 function Playground(): JSX.Element {
   const [state, setState] = useHashState({
-    jsx: false,
+    fileType: 'ts',
```
Comment (bradzacher) [inline packages/website/src/components/Playground.tsx:53]: I think we should default this to TSX.
IMO it's the rarer case that people want the non-TSX supported syntax (which really is just angle bracket assertions), and much more common that people want to use JSX in the playground.

File: packages/website/src/components/OptionsSelector.tsx [hunk 5]
```diff
@@ -74,11 +74,12 @@ function OptionsSelectorContent({
...
       <Expander label="Options">
-        <InputLabel name="Enable jsx">
-          <Checkbox
-            name="jsx"
-            checked={state.jsx}
-            onChange={(e): void => setState({ jsx: e })}
+        <InputLabel name="File type">
+          <Dropdown
+            name="fileType"
+            value={state.fileType ?? 'ts'}
```
Comment (bradzacher) [inline packages/website/src/components/OptionsSelector.tsx:80]: ditto below - can we default to TSX?

File: packages/website/src/components/types.ts (hunk 3 above)
Comment (armano2) [reply packages/website/src/components/types.ts:26]: my issue with using ts in this context is that we do not have access to it yet, we have to wait for  playground to be fully initialized, and to do so we have to know what extensions has to be choosen,
addtionally we will have to parse remove some options from it `TsBuildInfo` and `Json`

File: packages/website/src/components/editor/LoadedEditor.tsx (hunk 1 above)
Comment (armano2) [reply packages/website/src/components/editor/LoadedEditor.tsx:42]: in theory we could use it in here

```
export function determineLanguage(file: string): string {
 const kind = getScriptKind()
 switch (extension) { 
   case window.ts.ScriptKind.JS:
   
 }
}
```

but we still need extra check for `eslintrc`

File: packages/website/src/components/config.ts (hunk 2 above)
Comment (armano2) [reply packages/website/src/components/config.ts:20]: a tthis point we do not have access to typescript instance yet as editor may be not loaded

File: packages/website/src/components/editor/LoadedEditor.tsx (hunk 1 above)
Comment (armano2) [reply packages/website/src/components/editor/LoadedEditor.tsx:42]: after further considiration, i actually don't need it i can just let ts decide what language to choose base on available extensions

**Comment by armano2** [comment]
fallback cases to test:

- old url without jsx
  https://deploy-preview-6785--typescript-eslint.netlify.app/play/#ts=5.0.2&sourceType=module&showAST=ts&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA
- old url wiht jsx
  https://deploy-preview-6785--typescript-eslint.netlify.app/play/#ts=5.0.2&jsx=true&sourceType=module&showAST=ts&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

File: packages/website/src/components/hooks/useHashState.ts [hunk 6]
```diff
@@ -165,66 +172,33 @@ const writeStateToLocalStorage = (newState: ConfigModel): void => {
...
+      const url = `${window.location.pathname}#${newHash}`;
 
-  const _setState = useCallback((cfg: Partial<ConfigModel>) => {
-    console.info('[State] updating config diff', cfg);
-    setTmpState(cfg);
+      if (cfg.ts) {
+        window.location.href = url;
+        window.location.reload();
+      } else {
+        window.history.replaceState(undefined, document.title, url);
```
Comment (armano2) [inline packages/website/src/components/hooks/useHashState.ts:195]: note: this should be changed to use react router to fix issues with navigation 

but this is oo scope for this change

File: packages/website/src/components/hooks/useHashState.ts [hunk 7]
```diff
@@ -65,20 +73,22 @@ const parseStateFromUrl = (hash: string): ConfigModel | undefined => {
       );
     }
 
+    let fileType = readFileType(searchParams.get('fileType'));
+    if (searchParams.get('jsx') === 'true') {
+      fileType = '.tsx';
+    }
```
Comment (JoshuaKGoldberg) [inline packages/website/src/components/hooks/useHashState.ts:79]: [Nit] 

```suggestion
    const fileType =
      searchParams.get('jsx') === 'true'
        ? '.tsx'
        : readFileType(searchParams.get('fileType'));
```

**Review by JoshuaKGoldberg — APPROVED** [review]
🔥

**Pull request merged**